File: apps/pos_app/lib/ui/features/inventory/shrinkage/merma_taxonomy.dart
const mermaReasons = <String, String>{
  'VENCIDO': 'VENCIDO',
  'DESECHO_COCINA': 'DESECHO_COCINA',
  'DETERIORO_BODEGA': 'DETERIORO_BODEGA',
  'CORTESIA_DEGUSTACION': 'CORTESIA_DEGUSTACION',
};

const mermaReasonAliases = <String, String>{
  'VENCIMIENTO': 'VENCIDO',
  'MALA_PREPARACION': 'DESECHO_COCINA',
  'ROTO': 'DETERIORO_BODEGA',
  'DETERIORADO': 'DETERIORO_BODEGA',
  'CORTESIA': 'CORTESIA_DEGUSTACION',
};

const shrinkageTypes = <String>[
  'VENCIDO',
  'DESECHO_COCINA',
  'DETERIORO_BODEGA',
  'CORTESIA_DEGUSTACION',
];

String? normalizeMermaReason(String value) {
  final normalized = value.trim().toUpperCase();
  if (mermaReasons.containsKey(normalized)) {
    return normalized;
  }

  return mermaReasonAliases[normalized];
}

String requireMermaObservation(String value) {
  final observation = value.trim();
  if (observation.isEmpty) {
    throw ArgumentError('Merma observation is required');
  }

  return observation;
}

File: apps/admin_backend/src/modules/inventory/merma-taxonomy.ts
import { BadRequestException } from '@nestjs/common';

export const MERMA_REASONS = {
  VENCIDO: 'VENCIDO',
  DESECHO_COCINA: 'DESECHO_COCINA',
  DETERIORO_BODEGA: 'DETERIORO_BODEGA',
  CORTESIA_DEGUSTACION: 'CORTESIA_DEGUSTACION',
} as const;

export type MermaReason = (typeof MERMA_REASONS)[keyof typeof MERMA_REASONS];

const MERMA_REASON_ALIASES = {
  VENCIMIENTO: MERMA_REASONS.VENCIDO,
  MALA_PREPARACION: MERMA_REASONS.DESECHO_COCINA,
  ROTO: MERMA_REASONS.DETERIORO_BODEGA,
  DETERIORADO: MERMA_REASONS.DETERIORO_BODEGA,
  CORTESIA: MERMA_REASONS.CORTESIA_DEGUSTACION,
} as const satisfies Record<string, MermaReason>;

export function normalizeMermaReason(value: string): MermaReason | null {
  const normalized = value.trim().toUpperCase();
  const canonicalReasons = Object.values(MERMA_REASONS);
  if (canonicalReasons.includes(normalized as MermaReason)) {
    return normalized as MermaReason;
  }

  return (
    MERMA_REASON_ALIASES[normalized as keyof typeof MERMA_REASON_ALIASES] ??
    null
  );
}

export function requireMermaObservation(value: string): string {
  const observation = value.trim();
  if (observation.length === 0) {
    throw new BadRequestException('Merma observation is required');
  }

  return observation;
}
